### DISCUSSION TITLE
对导入/导出工具进行了优化

### DISCUSSION THREAD

**Discussion by limeng32** [apache/iotdb#4763] [Show and tell]
各位好。我最近在生产环境使用iotdb，发现导入工具在真实场景几乎无法使用，我研究了这一块代码，发现stream并没有被正确使用，于是我对其进行重构，同时我还增加了其它一些功能。我所做的事情如下：

1. 优化导入的实现，使得备份文件中的记录可以逐行读入内存、写入数据库、移出内存，新的工具对备份文件大小不再有任何要求；
2. 增加压缩的选项。在导出的同时进行压缩是常见行为，同时导入时也要支持从压缩文件导入，目前我针对gzip和snappy两种方式进行了实现；
3. 在导入/导出时有相应进度反馈。目前我采用的是每导出1千万点位或者导入1百万点位进行一次反馈，这样无论什么类型数据都可以有合适的反馈节奏；
4. 因为几乎完全没有使用原本代码，所以新工具名为export-csv-new和import-csv-new，和现有工具共存；

以上这些功能的pr我即将提交，如果各位有什么好想法也欢迎留言。

**Comment by limeng32** [comment]
我的pr请见：https://github.com/apache/iotdb/pull/4774

为了使用压缩功能，我在导出时新增一个可选参数“-c”，接受gizp和snappy参数，其它情况下默认为不压缩。同时在导入也有这样一个可选参数，如果使用的话由它指定解压方式，如果不指定的话通过备份文件后缀（.gz或.snappy）自动判断解压方式。见下图：
![image](https://user-images.githubusercontent.com/6504776/148875287-4a10de54-43ae-4f7c-b51b-5851957274ae.png)

**Comment by limeng32** [comment]
我注意到之前不是在master分支上提交pr，所以我重新提交了一个： https://github.com/apache/iotdb/pull/4774

**Comment by HTHou** [comment]
抱歉，我们本周在进行 0.13.0 版本发布准备工作，可能下周才可以review一下这个PR。欢迎社区所有人的代码、意见和建议！:>

**Comment by limeng32** [comment]
好的，各位辛苦！

**Comment by jixuan1989** [comment]
@limeng nice work! Could you provide a performance comparison?

**Comment by limeng32** [comment]
I will do it today.

**Comment by limeng32** [comment]
use 121151040 points to test export and import

export-tool-new:

- Plaintext: time cost 36.89s, about 3284100 p/s

- gzip: time cost 30.78s, about 3936000 p/s

- snappy: time cost 29.64s, about 4087400 p/s

import-tool-new:

- Plaintext: time cost 67.12s, about 1805000 p/s

- gzip: time cost 74.44s, about 1627500 p/s

- snappy: time cost 71.11s, about 1703700 p/s

export-tool:

- Plaintext: time cost 31.07s, about 3899300 p/s

import-tool:

- files bigger than 50MB cannot be imported

**Reply by jixuan1989** [reply]
Nice work!
